### PR TITLE
Hovertips for Feature Layers

### PR DESCRIPTION
**Pull request by james-rae**
## Description
Closes https://github.com/fgpv-vpgf/fgpv-vpgf/issues/1491

## Testing
Tested against `index-one` and `index-fgp`. Also against local custom config to test the ability to turn off hover tips on select layers.

## Documentation
JSDoc.
Config schema has changed, will require version bump when `v1.5` is finalized.

## Checklist

- [x] commits messages follow the guidelines
- [x] code passes unit tests
- [x] release notes have been updated
- [x] PR targets the correct release version

---
This change is [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1569)

### CODE REVIEW

**Comment by AleksueiR** [comment]
:lgtm:

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-:-KaDza0L_VQptNR_nqow:bnfp4nl)*

**Comment by AleksueiR** [comment]
Reviewed 3 of 6 files at r2.
Review status: 3 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/locales/translations.csv, line 172 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE-TfXp23Sk_OyrYum:-KaE-TfXp23Sk_OyrYun:b6r9rq7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/locales/translations.csv#L172)):*
> ```CSV
> ,toc.menu.visibility.allon,Show All,Montre tout
> ,toc.menu.visibility.alloff,Hide All,Cacher tout
> ,toc.menu.expand.title,Toggle Groups,Alternar grupos
> ```

I think you brought up an old version of translation file. These are incorrect (Spanish :)) translations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by james-rae** [comment]
*[src/locales/translations.csv, line 172 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE-TfXp23Sk_OyrYum:-KaE0Fj88kAaLr1wQxJS:bx9fp91) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/locales/translations.csv#L172)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

I think you brought up an old version of translation file. These are incorrect (Spanish :)) translations.
</blockquote></details>

Weird. I'll fix and re-commit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by AleksueiR** [comment]
Reviewed 2 of 6 files at r2.
Review status: 5 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by AleksueiR** [comment]
Reviewed 1 of 1 files at r3.
Review status: 5 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by alyec** [comment]
Commit comments got mangled somewhere in there.

---

Reviewed 5 of 7 files at r1, 5 of 6 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[src/app/geo/init-map.directive.js, line 65 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE2gAE7M7gBJog4ohX:-KaE2gAE7M7gBJog4ohY:be52sg6) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/init-map.directive.js#L65)):*
> ```JavaScript
>              * @function mouseDownHandler
>              * @private
>              */
> ```

missing jsdoc for param

---

*[src/app/geo/init-map.directive.js, line 77 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE2XdsE6pM6jvGR1J5:-KaE2XdsE6pM6jvGR1J6:bu29ofg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/init-map.directive.js#L77)):*
> ```JavaScript
>              * @function mouseMoveHandlerBuilder
>              * @private
>              */
> ```

missing jsdoc for params and return

---

*[src/app/geo/layer-record.class.js, line 490 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE4TxoDT2Bq3yT2icN:-KaE4TxoDT2Bq3yT2icO:bmb77s1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-record.class.js#L490)):*
> ```JavaScript
>                     // TODO after refactor, the class should have a .featureIdx property to use instead.
>                     const featureIdx = this.attributeBundle.indexes[0];
>                     this.attributeBundle[featureIdx].layerData.then(lInfo => {
> ```

Can the two levels of promises be flattened?

---

*[src/app/geo/layer-registry.service.js, line 460 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5THTcrXmZDU0djYn:-KaE5THTcrXmZDU0djYo:b5keikh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-registry.service.js#L460)):*
> ```JavaScript
>             /**
>              * Handles removal of a hovertip and cleanup tracking.
>              * @function destroyHovertip
> ```

`@private` ?

---

*[src/app/geo/layer-registry.service.js, line 470 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5EyXUSXR4KHkaobB:-KaE5EyXUSXR4KHkaobC:bbsk2mo) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-registry.service.js#L470)):*
> ```JavaScript
>             }
> 
>             // TODO find a better home for this function after grand refactor
> ```

Adding more random functions to layer registry :trophy: 

---

*[src/app/ui/tooltip/tooltip-hover.html, line 1 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5n3IKXA2el5um9wN:-KaE5n3IKXA2el5um9wO:b-eezzfm) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip-hover.html#L1)):*
> ```HTML
> 
> ```

typo tooltip (last instance)

---

*[src/app/ui/tooltip/tooltip.directive.js, line 9 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE6ODQ5M_75wbx_xLi:-KaE6ODQ5M_75wbx_xLj:b-vx0ijv) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.directive.js#L9)):*
> ```JavaScript
>      * @description
>      *
>      * The `rvTooltip` represents a tooltip on the map. Tooltip templates usually correspond with tooltip movement strategy.
> ```

usually?

---

*[src/app/ui/tooltip/tooltip.directive.js, line 40 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE6gH8Pjpv-1pXciO1:-KaE6gH8Pjpv-1pXciO2:b-ncjsq9) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.directive.js#L40)):*
> ```JavaScript
>                 }, 20, false, true);
> 
>                 scope.$watch(() => {
> ```

The indentation of the watch args is hard to follow, can you clean it up a bit?

---

*[src/app/ui/tooltip/tooltip.service.js, line 21 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE71Npux8XA3SM9sDH:-KaE71Npux8XA3SM9sDI:bpowlwe) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L21)):*
> ```JavaScript
>             /**
>              *
>              * Tooltip's origin point is generaly the position of the mouse cursor or clientX/Y of a mouse event.
> ```

typo generally

---

*[src/app/ui/tooltip/tooltip.service.js, line 24 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE7U28Y4K6RHLos5wJ:-KaE7U28Y4K6RHLos5wK:brvdixg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L24)):*
> ```JavaScript
>              *
>              * @function constructor
>              * @param {Object} movementStrategy specifies how the tooltip moves on the screen
> ```

Can you link to where the strategies are defined?

---

*[src/app/ui/tooltip/tooltip.service.js, line 26 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE7Y-ekoT_3a7Dkn-r:-KaE7Y-ekoT_3a7Dkn-s:b-hp1b1h) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L26)):*
> ```JavaScript
>              * @param {Object} movementStrategy specifies how the tooltip moves on the screen
>              * @param {Object} collisionStrategy specified how the tooltip reacts to collisions
>              * @param {String} content tooltips content
> ```

Is this an HTML string?

---

*[src/app/ui/tooltip/tooltip.service.js, line 87 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE8cozaLXgNjh2Vksi:-KaE8cozaLXgNjh2Vksj:b-2zg97w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L87)):*
> ```JavaScript
> 
>             /**
>              * Returns origin point (mouse point) of the tooltip relative to its parent container.
> ```

I'm not following exactly what the parent container is.

---

*[src/app/ui/tooltip/tooltip.service.js, line 91 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE8Atn2DqBKpQqDbw7:-KaE8Atn2DqBKpQqDbw8:b-czbjov) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L91)):*
> ```JavaScript
>              * @function getOriginPoint
>              * @param {Boolean} includeRunningOffset if set, returns tooltip origin point including the running offset
>              */
> ```

Missing `@return`

---

*[src/app/ui/tooltip/tooltip.service.js, line 177 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE9DDgq9qoyEJrxxcB:-KaE9DDgq9qoyEJrxxcC:brwq43j) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L177)):*
> ```JavaScript
>                 }
> 
>                 this.offset(0, 0);
> ```

This looks like a noop but I assume that offset has some useful side effect which is why it gets called like this?

---

*[src/app/ui/tooltip/tooltip.service.js, line 203 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE9T1pa6BkokZPBZuK:-KaE9T1pa6BkokZPBZuL:b5alqqi) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L203)):*
> ```JavaScript
> 
>             /**
>              * Checks if there is any collision between teh supplied item and the target container. Returns a vector to prevent collision.
> ```

typo: the (first instance)

---

*[src/app/ui/tooltip/tooltip.service.js, line 228 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEAIl2yXIk_ess-nnC:-KaEAIl2yXIk_ess-nnD:biv0dot) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L228)):*
> ```JavaScript
>         }
> 
>         // movementStrategy
> ```

In the case of these strategies I think the jsdoc describing their behaviour should be be moved to the class level (I think jsdoc parses it properly) and the constructor comments can be minimal.

---

*[src/app/ui/tooltip/tooltip.service.js, line 387 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEBQeR-GRz8-mLLKCH:-KaEBQeR-GRz8-mLLKCI:bwiz51d) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L387)):*
> ```JavaScript
>             ref.containInsideStrategy = new ContainInside(storageService.panels.shell);
> 
>             // test code, remove before deployment //
> ```

If you're done with this can we clean it up before merging?

---

*[src/app/ui/tooltip/tooltip.service.js, line 448 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEBHdGz-Z7gxcYdBxs:-KaEBHdGz-Z7gxcYdBxt:b-738573) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L448)):*
> ```JavaScript
>          * @param {String} content tooltip content
>          * @param {Object} self a self object that will be available on the tooltip directive scope
>          */
> ```

missing `@return`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-:-KaECABfLQRUhjmZM-IN:b-80l4p2)*

**Comment by james-rae** [comment]
I warned him about making me use git commands outside of my normal flow. Warnings were ignored.

---

Review status: 11 of 12 files reviewed at latest revision, 18 unresolved discussions.

---

*[src/app/geo/layer-record.class.js, line 490 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE4TxoDT2Bq3yT2icN:-KaEMmwmZ1xLG5jGo2ZI:b-std9qg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-record.class.js#L490)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Can the two levels of promises be flattened?
</blockquote></details>

Was split to facilitate enhancements in https://github.com/fgpv-vpgf/fgpv-vpgf/issues/1571 . If that gets implemented, we may need to choose what method gets called for the second asynch request to get attribtues.

---

*[src/app/geo/layer-registry.service.js, line 460 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5THTcrXmZDU0djYn:-KaENDcY0u0Rr0R99MgZ:b-mhsq5s) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-registry.service.js#L460)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

`@private` ?
</blockquote></details>

donethanks

---

*[src/app/geo/layer-registry.service.js, line 470 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5EyXUSXR4KHkaobB:-KaENEx9mM-9mUECoOrN:by8epng) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-registry.service.js#L470)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Adding more random functions to layer registry :trophy: 
</blockquote></details>

The beauty is most of the other functions i need are already here. I propose we migrate to a 1-file architecture.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-:-KaENyOLor2tX8mTRd9i:b-331wtv)*

**Comment by alyec** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[src/app/geo/layer-registry.service.js, line 470 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5EyXUSXR4KHkaobB:-KaHmMevBLQc9q5wGV3n:b-f8r5ev) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-registry.service.js#L470)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

The beauty is most of the other functions i need are already here. I propose we migrate to a 1-file architecture.
</blockquote></details>

Sounds reasonable.  Namespaces are overrated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by AleksueiR** [comment]
Review status: 9 of 13 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[src/app/ui/tooltip/tooltip-hover.html, line 1 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE5n3IKXA2el5um9wN:-KaHbbCeNlVHGUfurRJ_:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip-hover.html#L1)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

typo tooltip (last instance)
</blockquote></details>

Done.

---

*[src/app/ui/tooltip/tooltip.directive.js, line 9 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE6ODQ5M_75wbx_xLi:-KaHcGQp9gkSeh7_oUaa:bsj646i) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.directive.js#L9)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

usually?
</blockquote></details>

Corrected.

---

*[src/app/ui/tooltip/tooltip.directive.js, line 40 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE6gH8Pjpv-1pXciO1:-KaHdnGXIZyA4sY3_zQ9:begbe5c) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.directive.js#L40)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

The indentation of the watch args is hard to follow, can you clean it up a bit?
</blockquote></details>

Any better now?

---

*[src/app/ui/tooltip/tooltip.service.js, line 21 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE71Npux8XA3SM9sDH:-KaHe3jYK_ZKGobY8b67:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L21)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

typo generally
</blockquote></details>

Done.

---

*[src/app/ui/tooltip/tooltip.service.js, line 24 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE7U28Y4K6RHLos5wJ:-KaHeogNPixnIUEgdwjy:b-vvla7j) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L24)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Can you link to where the strategies are defined?
</blockquote></details>

Good enough?

---

*[src/app/ui/tooltip/tooltip.service.js, line 26 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE7Y-ekoT_3a7Dkn-r:-KaHfGYPZmR-ATBYQmOL:bgwt35a) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L26)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Is this an HTML string?
</blockquote></details>

Yes. Fixed.

---

*[src/app/ui/tooltip/tooltip.service.js, line 87 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE8cozaLXgNjh2Vksi:-KaHfc--VLpTSsYdYsVy:bcmh6r7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L87)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

I'm not following exactly what the parent container is.
</blockquote></details>

Added note to the constructor comment block.

---

*[src/app/ui/tooltip/tooltip.service.js, line 91 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE8Atn2DqBKpQqDbw7:-KaHgIcP2y2yhpyYKvlE:bt2vr1b) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L91)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Missing `@return`
</blockquote></details>

Fixed.

---

*[src/app/ui/tooltip/tooltip.service.js, line 177 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE9DDgq9qoyEJrxxcB:-KaHhx-ndhwEd5vQH6Mi:b-vyrl4w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L177)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

This looks like a noop but I assume that offset has some useful side effect which is why it gets called like this?
</blockquote></details>

`offset` re-applies the current running offset and checks for collisions.

---

*[src/app/ui/tooltip/tooltip.service.js, line 203 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE9T1pa6BkokZPBZuK:-KaHiAHLUS_MfDa2jBXb:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L203)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

typo: the (first instance)
</blockquote></details>

Done.

---

*[src/app/ui/tooltip/tooltip.service.js, line 228 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEAIl2yXIk_ess-nnC:-KaHijWzsNOmmRqsELl4:b-bjpj7a) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L228)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

In the case of these strategies I think the jsdoc describing their behaviour should be be moved to the class level (I think jsdoc parses it properly) and the constructor comments can be minimal.
</blockquote></details>

Better?

---

*[src/app/ui/tooltip/tooltip.service.js, line 387 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEBQeR-GRz8-mLLKCH:-KaHl7_th3XomGML82uI:bss4a2b) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L387)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

If you're done with this can we clean it up before merging?
</blockquote></details>

Removed.

---

*[src/app/ui/tooltip/tooltip.service.js, line 448 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaEBHdGz-Z7gxcYdBxs:-KaHlDYM5scjHMOksORk:b-896fix) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/ui/tooltip/tooltip.service.js#L448)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

missing `@return`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by alyec** [comment]
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by james-rae** [comment]
Review status: 11 of 13 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/app/geo/layer-record.class.js, line 490 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569#-KaE4TxoDT2Bq3yT2icN:-KaJABoi4CsaZ7iCKu5u:bikmekp) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/aaa4937de514e11170dec384ef7cf3785447ae21/src/app/geo/layer-record.class.js#L490)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

Was split to facilitate enhancements in https://github.com/fgpv-vpgf/fgpv-vpgf/issues/1571 . If that gets implemented, we may need to choose what method gets called for the second asynch request to get attribtues.
</blockquote></details>

Flattenified to a structure that can adapt to future enhancements

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by alyec** [comment]
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by AleksueiR** [comment]
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1569)*

**Comment by alyec** [comment]
Cleaned up version in #1584 closing this one
